### ISSUE TITLE
GM_Cookie

### ISSUE THREAD

**Issue by qsniyg** [violentmonkey/violentmonkey#603]
Implementation of https://github.com/Tampermonkey/tampermonkey/issues/465 and https://github.com/greasemonkey/greasemonkey/issues/1802 . This would be very useful for my userscript, as it requires a few API calls that need cookies from the websites to access otherwise private user data (original images for Flickr etc.). `document.cookie` does not contain server-side cookies, which are needed for this unfortunately.

**Comment by tophf** [comment]
GM_Cookie is not supported by other userscript managers (TM only has it in beta) so we won't implement it either. Violentmonkey now includes protected HTTP headers from GM_xmlHttpRequest so it should be possible to read all of the cookies from `Set-Cookie` header in [response.responseHeaders](https://violentmonkey.github.io/api/gm/#gm_xmlhttprequest) multi-line string.

**Comment by qsniyg** [comment]
Would it be possible to reconsider this feature? There are so many cases now where I've been unable to support websites because of lack of `GM_Cookie` support in VM (e.g. due to CSRF protection, requiring a header to match a cookie). `Set-Cookie` is only helpful if the web server returns it consistently, which unfortunately is rarely the case. If I were to try to force it to return the Set-Cookie header by removing a field from the Cookie header, it will overwrite the user's cookies, resulting in the user possibly being logged out. Yes it's annoying and (imo) stupid on the part of the server, but unfortunately there's no way I can see to go around this limitation.

**Comment by tophf** [comment]
I guess we can implement it once it's out of beta in Tampermonkey.

**Comment by Neurotoxin001** [comment]
So I use Tampermonkey 4.10 and it's not a beta (https://www.tampermonkey.net/changelog.php?ext=gcal&show=dhdg)
And GM_Cookie is supported here (because there is no this issue https://github.com/violentmonkey/violentmonkey/issues/1024 with Tampermonkey).

You still not planing to implement GM_Cookie to VM?

**Comment by tophf** [comment]
No. I've just debugged Tampermonkey stable and it does not actually implement GM_cookie. All it has is a dummy function that always returns an `undefined` result and `not supported` in the second parameter (this is the error). Try asking the userscript author to use a dummy polyfill so that their script will be able to run in more engines even if it's not in fully functional mode.

**Comment by Neurotoxin001** [comment]
But why can you add same dummy function to VM?

**Comment by tophf** [comment]
We will implement it once it's out of beta in Tampermonkey.

**Comment by JenieX** [comment]
I'm thinking of creating an npm package, something similar to this [instagram-stories](https://github.com/jlobos/instagram-stories).

Similar packages and scrappers usually require credentials (cookies). Most user scripts developers would go through the trouble of creating a tutorial on how to extract the required credentials, [see this developer](https://greasyfork.org/en/scripts/372366-instagram-source-opener). That is why I think it would be a good idea for me to create a user script that exports the required credentials to a file that later can be used with my npm package.

**Comment by KaKi87** [comment]
Hello @tophf & @gera2ld, I encountered an issue that only this API can solve, please consider implementing it.

I reported this at 3 different places, choose where to read it from :

- https://github.com/Tampermonkey/tampermonkey/issues/465#issuecomment-1507712230
- https://reddit.com/r/userscripts/comments/12l5af9/weird_cookie_behavior/
- https://discord.com/channels/995346102003965952/1094107789980348488/1094107789980348488

Thanks

**Comment by KaKi87** [comment]
up

**Comment by tophf** [comment]
> it's out of beta for a long while now, like 2 years?

It's not working outside of beta and just prints "not supported" in console.

**Comment by tophf** [comment]
### We will implement it once it's out of beta in Tampermonkey
# and actually works.

**Comment by tophf** [comment]
TM enabled GM_cookie in its stable version, so I guess it means we can add it too.

**Comment by n00b42** [comment]
> TM enabled GM_cookie in its stable version, so I guess it means we can add it too except for HttpOnly (TM beta only).

According to https://github.com/Tampermonkey/tampermonkey/issues/465#issuecomment-2479061865 HttpOnly is in stable as well (but needs to be enabled in settings)

**Comment by KaKi87** [comment]
@tophf 🙏

**Comment by kittizz** [comment]
GM_cookie implement plzzz 🙏🙏

**Comment by piknockyou** [comment]
I'm really looking forward to this feature! 

I have a userscript that generates batch files for downloading private videos via yt-dlp from platforms like Instagram and other video sites. The script automatically captures the video URL and creates a ready-to-run batch file - it's a two-click workflow.

However, many platforms (especially Instagram private profiles and authenticated video sites) require cookies for yt-dlp to work. Currently, I have to manually use the "Get cookies.txt LOCALLY" browser extension every time: open the extension menu, click export, save the cookies.txt file, then run my batch file.

With GM.cookie API support, my userscript could automatically export the cookies alongside the batch file in one step - making it a truly seamless two-click process instead of the current multi-step workflow. This would eliminate the need for a separate extension and manual cookie export each time I want to download a video.

HttpOnly cookie access is essential for this use case since authentication cookies are typically HttpOnly for security reasons.

**Comment by KaKi87** [comment]
Also, all the screenshots and issues are in Chinese.
